rosplan_action_interface/src/ServiceActionInterface.py

```diff
@@ -0,0 +1,109 @@
+        self._has_default_msg_type = ("default_service_type" in action_config)
+        self._has_default_request = ("default_service_request" in action_config)
+
+    def run(self, dispatch_msg):
+        # Run on a new thread
+        new_thread = threading.Thread(target=self.run_thread, args=(dispatch_msg,))
+        new_thread.start()
+
+    def run_thread(self, dispatch_msg):
+
+        rospy.loginfo('KCL: ({}) Interfacing service {}'.format(rospy.get_name(), self._action_name))
+
+        # set action status
+        plan_action_id = (dispatch_msg.plan_id, dispatch_msg.action_id)
+        self._action_dispatch_msg[plan_action_id] = dispatch_msg
+        self._action_status[plan_action_id] = ActionFeedback.ACTION_ENABLED
+
+        # load default configuration for request
+        if self._has_default_topic:
+            action_server = self.parse_config_string(self._action_config["default_service"], dispatch_msg)[0]
+
+        if self._has_default_msg_type:
+            goal_msg_type = self.parse_config_string(self._action_config["default_service_type"], dispatch_msg)[0]
+
+        override_request = None
+        override_result = None
+        if "pddl_parameters" in self._action_config:
+            pddl_params = self._action_config["pddl_parameters"]
+            for config in self._action_config["parameter_values"]:
+                # Check parameter is correct
+                if self.params_match(config, pddl_params, dispatch_msg):
+                    if "service" in config:
+                        action_server = self.parse_config_string(config["service"], dispatch_msg)[0]
+                    if "service_type" in config:
+                        goal_msg_type = self.parse_config_string(config["service_type"], dispatch_msg)[0]
+                    if "service_request" in config:
+                        override_request = config["service_request"]
+                    if "service_result" in config:
+                        override_result = config["service_result"]
+
+        # import goal msg type
+        i = goal_msg_type.find('/')
+        pkg_name = goal_msg_type[:i]
+        msg_name = goal_msg_type[i+1:]
+
+        pkg = importlib.import_module(pkg_name+".srv")
+        srv = getattr(pkg,msg_name)
+        req = getattr(pkg,msg_name+"Request")
+
+        # create client and goal msg
+        srv_proxy = rospy.ServiceProxy(action_server, srv)
+        rospy.wait_for_service(action_server, timeout=10)
+        request = req()
+
+        # populate goal msg
+        if self._has_default_request:
+            for param in self._action_config["default_service_request"]:
+                value = self._action_config["default_service_request"][param]
+                self.populate_goal_msg(request, param, value, dispatch_msg)
+
+        if override_request:
+            for param in override_request:
+                value = override_request[param]
+                self.populate_goal_msg(request, param, value, dispatch_msg)
+
```

Comment:
Would it be better to have the structure
if self._has_default_request:
 ...
elif override_request:
 ...
else:
  rospy.warn(...)
? 
For the case that no configuration is defined for that dispatch msg?

Reply:
I did something like this while fixing the comment above:
```
if self._has_Default_request:
  # check each param, unless overriden
if override_response is not None:
  # check each param
```

Reply:
First: once you have request and once response. Is this only a typo?

Second: We can have the case that no default is defined and let's say that options only for parameter values p1 and p2 are set up. If a dispatch msg comes with a parameter value p3, no warning is generated and probably an error will be thrown at some point in the execution

Reply:
(First) Ah, in my comment above, yes it is a typo. In the code request and response should all be in the righ places.

Reply:
(Second) This applies only to the request and actionlib goal - in the case of result/response the behaviour should work fine. The result is true only if the parameters set in the override all match, and other non-overidden parameters match in the default. The two groups can be subsets of all possible parameters, overlap or not, and be empty or not. If both are empty then the result is always true.

For the request, I'll open it up again. We should have a similar logic:
```
# set all default parameters that are specified (could be subset)
# set overrides for the action params
```
About printing a warning - not sure how we could do this. We could read the action goal message and print a warning unless every single field was set - but much of the time there are optional fields. I think it is up to the actionlib/service provider to print an error/warning about what was missing.

Reply:
Yes, I got it. I had a mistake in my logic.